org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/references/search/SearchQueryFactory.java

```diff
@@ -0,0 +1,261 @@
+/*******************************************************************************
+* Copyright (c) 2023 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.extensions.references.search;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.eclipse.lemminx.dom.DOMAttr;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.dom.DOMElement;
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.dom.DOMText;
+import org.eclipse.lemminx.extensions.references.search.SearchNode.Direction;
+import org.eclipse.lemminx.extensions.references.search.SearchQuery.QueryDirection;
+import org.eclipse.lemminx.extensions.references.settings.XMLReferenceExpression;
+import org.eclipse.lemminx.extensions.references.settings.XMLReferences;
+import org.eclipse.lemminx.extensions.references.settings.XMLReferencesSettings;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+/**
+ * XML references search Query factory.
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class SearchQueryFactory {
+
+	public static Direction matchNode(DOMNode node, XMLReferenceExpression expression, QueryDirection direction) {
+		switch (direction) {
+			case FROM_2_TO:
+				if (expression.matchTo(node)) {
+					return Direction.TO;
+				}
+				return Direction.NONE;
+			case TO_2_FROM:
+				if (expression.matchFrom(node)) {
+					return Direction.FROM;
+				}
+				return Direction.NONE;
+			default:
+				if (expression.matchFrom(node)) {
+					return Direction.FROM;
+				}
+				if (expression.matchTo(node)) {
+					return Direction.TO;
+				}
+				return Direction.NONE;
+		}
+	}
+
+	public static Direction matchNode2(DOMNode node, XMLReferenceExpression expression, QueryDirection direction) {

```

Comment:
Add a Javadoc to these methods explaining how they are different.

Whenever I see `methodName()` and `methodName2()` in JDT/JDT-LS code, `methodName()` is usually deprecated and `methodName2()` is a method with the same name that does the same thing but with bugs fixed. Since it's not the case here, I think it's important to distinguish these methods, maybe by renaming them.

Reply:
fixed